slidingwindow.go

```diff
@@ -152,6 +153,70 @@ func (s *SlidingWindowRedis) Increment(ctx context.Context, prev, curr time.Time
 	}
 }
 
+// SlidingWindowMemcached implements SlidingWindow in Memcached.
+type SlidingWindowMemcached struct {
+	cli    *memcache.Client
+	prefix string
+}
+
+// NewSlidingWindowMemcached creates a new instance of SlidingWindowMemcached.
+func NewSlidingWindowMemcached(cli *memcache.Client, prefix string) *SlidingWindowMemcached {
+	return &SlidingWindowMemcached{cli: cli, prefix: prefix + ":SlidingWindow"}
+}
+
+// Increment increments the current window's counter in Memcached and returns the number of requests in the previous window
+// and the current one.
+func (s *SlidingWindowMemcached) Increment(ctx context.Context, prev, curr time.Time, ttl time.Duration) (int64, int64, error) {
+	var prevCount uint64
+	var currCount uint64
+	var err error
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+
+		var item *memcache.Item
+		prevKey := fmt.Sprintf("%s:%d", s.prefix, prev.UnixNano())
+		item, err = s.cli.Get(prevKey)
+		if err != nil {
+			if errors.Is(err, memcache.ErrCacheMiss) {
+				err = nil
+				prevCount = 0
+			} else {
+				return
+			}
+		} else {
+			prevCount, err = strconv.ParseUint(string(item.Value), 10, 64)
+			if err != nil {
+				return
+			}
+		}
+
+		currKey := fmt.Sprintf("%s:%d", s.prefix, curr.UnixNano())
+		currCount, err = s.cli.Increment(currKey, 1)
+		if err != nil && errors.Is(err, memcache.ErrCacheMiss) {
+			currCount = 1
+			item = &memcache.Item{
+				Key:   currKey,
+				Value: []byte(strconv.FormatUint(currCount, 10)),
+			}
+			err = s.cli.Add(item)
+		}
+	}()
+
+	select {
+	case <-done:
+		if err != nil {
+			if errors.Is(err, memcache.ErrNotStored) {
+				return s.Increment(ctx, prev, curr, ttl)
```

Comment:
It seems like we should return an error in this case and let the client decide how to handle it: otherwise we may end up in a loop

Reply:
The recursive Increment is trying to solve a race condition.

If two callers both `Increment` (line 195) at the same time while the counter doesn't exist, they both fail with `ErrCacheMiss` and `Add` (line 202) to initiate counter with value 1. Only one of them will succeed and the other will fail with `ErrNotStored`. The failing one can recursively call itself and do `Increment` (line 195) again, and the second time it should work as the counter is created already.

If you feel strongly against this implementation I can return error instead, otherwise I prefer to keep it as is.

Reply:
Could you please provide more details about this:

> Only one of them will succeed and the other will fail with ErrNotStored.

How is it guaranteed that at most 1 client will be able to add the key to memcached?